File: backend/Backend/src/main/java/com/sgw/backend/entity/VoucherListing.java
package com.sgw.backend.entity;

import com.fasterxml.jackson.annotation.JsonIgnore;
import com.sgw.backend.entity_venue.Venue;
import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.math.BigDecimal;
import java.time.LocalDate;
import java.util.List;

@Entity
@Data
@NoArgsConstructor
@AllArgsConstructor
public class VoucherListing {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private long voucherListingId;

    private LocalDate voucherListingDelistDate;

    private Integer validityPeriodInDays;

    private String description;

    private BigDecimal voucherValue;

    private BigDecimal voucherCost;

    private Long voucherStock;

    private boolean enabled;

    private String voucherName;

    private boolean adminBanned;

    private boolean completed;

    @JsonIgnore
    @ManyToOne(optional = false)
    private Owner owner;

    @OneToMany(cascade = CascadeType.ALL, orphanRemoval = false)
    private List<Voucher> vouchers;
}

File: student_frontend/lib/model/voucher_listing.dart
import 'voucher.dart';

class VoucherListing {
  final int voucherListingId;
  final String voucherListingDelistDate;
  final int validityPeriodInDays;
  final String description;
  final double voucherValue;
  final double voucherCost;
  final int voucherStock;
  final bool enabled;
  final String voucherName;
  final List<int> voucherIds;

  VoucherListing({
    required this.voucherListingId,
    required this.voucherListingDelistDate,
    required this.validityPeriodInDays,
    required this.description,
    required this.voucherValue,
    required this.voucherCost,
    required this.voucherStock,
    required this.enabled,
    required this.voucherName,
    required this.voucherIds,
  });

  factory VoucherListing.fromJson(Map<String, dynamic> json) {
    var voucherList = json['vouchers'] as List;
    List<Voucher> vouchers =
        voucherList.map((i) => Voucher.fromJson(i)).toList();
    List<int> voucherIds =
        vouchers.map((voucher) => voucher.billableId).toList();

    return VoucherListing(
      voucherListingId: json['voucherListingId'],
      voucherListingDelistDate: json['voucherListingDelistDate'],
      validityPeriodInDays: json['validityPeriodInDays'],
      description: json['description'],
      voucherValue: json['voucherValue'].toDouble(),
      voucherCost: json['voucherCost'].toDouble(),
      voucherStock: json['voucherStock'],
      enabled: json['enabled'],
      voucherName: json['voucherName'],
      voucherIds: voucherIds,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'voucherListingId': voucherListingId,
      'voucherListingDelistDate': voucherListingDelistDate,
      'validityPeriodInDays': validityPeriodInDays,
      'description': description,
      'voucherValue': voucherValue,
      'voucherCost': voucherCost,
      'voucherStock': voucherStock,
      'enabled': enabled,
      'voucherName': voucherName,
      'voucherIds': voucherIds,
    };
  }

  @override
  String toString() {
    return '''
VoucherListing(
  voucherListingId: $voucherListingId,
  voucherListingDelistDate: $voucherListingDelistDate,
  validityPeriodInDays: $validityPeriodInDays,
  description: $description,
  voucherValue: $voucherValue,
  voucherCost: $voucherCost,
  voucherStock: $voucherStock,
  enabled: $enabled,
  voucherName: $voucherName
)''';
  }
}
